File: race_officer/lib/features/results/domain/race_scorer.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:sailbrowser_flutter/features/race/domain/race_competitor.dart';
import 'package:sailbrowser_flutter/features/results/scoring/rating_system.dart';

import '../scoring/result_code_scoring.dart';

/// Functions related to calculating times and positions for an individual race.
class RaceScorer {
  RaceScorer();

  /// Calculates the corrected time taking into account any time penalty.
  /// The number of laps is scaled up to the maximum number of laps
  ({Duration corrected, Duration elapsed}) calculateResultTimes(
      RaceCompetitor res,
      RatingSystem scheme,
      bool isAverageLap,
      DateTime startTime,
      int maxLaps) {
    var corrected = 0.0;
    var elapsed = 0.0;

    if (res.finishTime != null) {
      final diff = res.finishTime!.difference(startTime).inSeconds.toDouble();

      // If start time is before finish time return defaults of zero
      if (diff < 0) {
        return (corrected: const Duration(), elapsed: const Duration());
      }

      elapsed = isAverageLap ? diff / res.numLaps * maxLaps : diff;

      switch (scheme) {
        case RatingSystem.py:
          corrected = elapsed * 1000 / res.handicap;
        case RatingSystem.irc:
          corrected = elapsed / res.handicap;
        default:
      }
    }

    return (
      // Round to nearest second rather than integer part
      corrected: Duration(seconds: corrected.round()),
      elapsed: Duration(seconds: elapsed.round())
    );
  }

  /// Return the number of starters in race taking into account result code
  int startersInRace(List<RaceCompetitor> results) {
    final res = results.fold<int>(0, (count, comp) {
      final code = getScoringData(comp.resultCode);
      return (code!.isStartAreaComp) ? count = count + 1 : count;
    });
    return res;
  }

    /// Return the number of finishers in race taking into account result code
  int finishersInRace(List<RaceCompetitor> results) {
    final res = results.fold<int>(0, (count, comp) {
      final code = getScoringData(comp.resultCode);
      return (code!.isFinishedComp) ? count = count + 1 : count;
    });
    return res;
  }

  /// Sort by corrected time taking into account resultCode
  int sortByCorrectedTime(RaceCompetitor a, b) {
    if (a.isOk && b.isOk) {
      return (a.correctedTime.inSeconds - b.correctedTime.inSeconds).toInt();
    } else if (a.isOk && !b.isOk) {
      return 1;
    } else if (!a.isOk && b.isOk) {
      return -1;
    } else {
      // a and b both not OK - keep sort order
      return 0;
    }
  }
/*
/// Assign posiiton and points for finishers
/// Competitors with the same ellapsed time get average of the two positions 
/// 
/// Time penalities should be applied befre calling this function
assignPointsForFinishers(results: Result[]) {
  let position = 1;
  let count = 1;
  let lasttime;
  for (const res of results) {
    if (lasttime !== res.elapsedTime) {
      position = count;
    }
    res.position = position;
    res.points = position;

    count = count + 1;
    lasttime = res.elapsedTime;
  }
}

assignPointsForNonFinishers(results: Result[], competitorsInSeries: number, shortSeries: boolean) {
  const starters = startersInRace(results);

  for (const res of results) {
    const resultCode = getScoringData(res.resultCode);

    const code = shortSeries ? resultCode.shortSeries : resultCode.longSeries;
    const factor = shortSeries ? resultCode.shortSeriesFactor : resultCode.longSeriesFactor;

    // Assign the times
    switch (code) {
      case 'StartArea':
        res.points = starters + factor;
      case 'scoringPenalty':
           // Scoring penalties are applied after all other positions have been determined.

        break;
      case 'InSeries':
        res.points = competitorsInSeries + factor;
      case 'AvgBefore':
          // 
          throw new Error("To Do");
        break;
      case 'AvgAll':
            throw new Error("To Do");

        break;
      case 'SetByHand':
            throw new Error("To Do");
        break;
      case 'NA':
        //  NA is either NoFinished or OK
        break;
    }

    switch
    if (resultCode.shortSeries === 'StartArea') {
      res.points = starters + resultCode.shortSeriesFactor;
    } else if (resultCode.shortSeries === 'PositionPenalty') {

    } else if

    // 'NA' | 'InSeries' | | 'AvgAll' | 'AvgBefore' | 'TimePenalty' |  | 'SetByHand';

  }
}

/// Scoring penalty appliey.  This is 20% for zfp and default of 20% for scp
/// This is applied after the points and order have been calculated.  
applyScoringPenalties() {
  // 
}

/** Calculates positions for a single race */
export function calculateRacePositions(results: Result[], race: Race, fleet: Fleet): Result[] {
  // Get the fleet of the races for the handicap scheme

  const updatedResults = results.map(res => {

    const times = calculateResultTimes(res, fleet.handicapScheme, race.isAverageLap, race.actualStart);

    res.correctedTime = times.corrected;
    res.elapsedTime = times.elapsed;
    return res;
  });

  // Assign points for race.
  results.sort((a, b) => sortByCorrectedTime(a, b));
  assignPointsForFinishers(results);
  assignPointsForNonFinishers(results);
  sortByPoints(results);
  applyScoringPenaties(results);
  handSetOverrides(results);
  sortByPoints(results);

  return updatedResults;

}

function sortByPoints(a: Result) {

}

 */
}

final raceScorerProvider = Provider((ref) => RaceScorer());

File: sailbrowser-web/src/app/scoring/services/race-scorer.ts
import { PublishedRace, RaceResult } from 'app/published-results/model/published-race';
import { Race, RaceType } from 'app/race-calender';
import { RaceCompetitor } from 'app/results-input';
import { HandicapSystem } from 'app/scoring/model/handicap-system';
import { getLongAlgorithm, getShortAlgorithm, isFinishedComp, isStartAreaComp, ResultCodeAlgorithm } from 'app/scoring/model/result-code-scoring';
import { SailbrowserError } from 'app/shared/utils/sailbrowser-error';
import { differenceInSeconds } from 'date-fns';
import { SeriesScoringScheme } from '../model/scoring-algotirhm';

/**
 * Uses competitor start, finish. lap and status to calculate results for a single race. 
 * All data in the ResultsData object is populated (elapsed/corrected times points and positions).
 * 
 * Assumes all competitors in the race are supplied.  
 * The maximum number of laps and number of starters is calculated based on all competitors. 
 */
export function scoreRace(
  race: Race,
  competitors: RaceCompetitor[],
  scheme: HandicapSystem,
  seriesType: SeriesScoringScheme,
  seriesCompetitorCount: number,
): RaceResult[] {

  // Build initial results from competitors.
  const results = buildRaceResults(competitors);

  // Calculate elapsed and corrected times for all competitors.
  calculateTimes(results, competitors, race.isAverageLap, scheme);

  // Determine the ordering property for finishers and sort them
  const orderingProperty = determineOrdering(race.type, scheme, results);
  // Sort all results to establish finishing order (finishers first, then by ordering property).
  results.sort((a, b) => sortByFinishingOrder(a, b, orderingProperty));

  // Assign points to finishers based on their sorted order.
  assignPointsForFinishers(results, orderingProperty, seriesType, seriesCompetitorCount);

  // Assign points for non-finishers 
  // Handless codes NOT based on series results (DNF, OCS, etc.).
  applyStaticRacePenalties(results, seriesCompetitorCount, seriesType);

  // Sort all competitors by points to determine final race ranks.
  results.sort((a, b) => sortByPoints(a, b));

  // Calcualte final ranks within race
  calculateRanks(results);

  return results;
}

/** 
 * Calculates competitor ranks within the race. 
 * Competitors tied on points will recieve the same rank.
 */
export function calculateRanks(results: RaceResult[]) {
  for (let i = 0; i < results.length; i++) {
    if (i > 0 && results[i].points === results[i - 1].points) {
      // Tied points, so assign the same rank
      results[i].rank = results[i - 1].rank;
    } else {
      // Not tied, or first competitor, so rank is position in the sorted list
      results[i].rank = i + 1;
    }
  }
}

/**
 * Determines the property to sort finishers 
 * and validates that all finishers have the required data.
 * Throws SailbrowserError if complet data to rank all competitors is not avaliable
 * Algorithm is: 
 * 1. If Pursuit -> position.
 * 2. If Level Rating -> if manualPositions -> position, else elapsedTime.
 * 3. Handicap -> correctedTime.
 */
function determineOrdering(raceType: RaceType, scheme: HandicapSystem, results: RaceResult[]): keyof RaceResult {
  let orderingProperty: keyof RaceResult;
  const finishers = results.filter(res => isFinishedComp(res.resultCode));

  if (raceType === 'Pursuit') {
    orderingProperty = 'rank';
    validateFinishersHaveData(finishers, 'rank', 'Pursuit races require a manual position');
  } else if (scheme === 'Level Rating') {
    const useManualPositions = finishers.some(f => f.rank > 0);
    if (useManualPositions) {
      orderingProperty = 'rank';
      validateFinishersHaveData(finishers, 'rank', 'Manual positions are used');
    } else {
      orderingProperty = 'elapsedTime';
      validateFinishersHaveData(finishers, 'elapsedTime', 'Finish times are used');
    }
  } else { // Handicap race
    orderingProperty = 'correctedTime';
    // For handicap races, a corrected time of 0 implies missing finish time, which is an error for a finisher.
    validateFinishersHaveData(finishers, 'correctedTime', 'Handicap races require a finish time');
  }
  return orderingProperty;
}

/**
 * Checks that all finishers have a valid (non-zero) value for the specified ordering property.
 * Throws a SailbrowserError if any finisher is missing data.
 */
function validateFinishersHaveData(finishers: RaceResult[], property: keyof RaceResult, context: string) {
  const missingData = finishers.find(f => !((f[property] as number) > 0));
  if (missingData) {
    const propertyName = property === 'rank' ? 'position' : 'finish time';
    throw new SailbrowserError(`Inconsistent ordering data: ${context}, but finisher with sail number ${missingData.sailNumber} is missing a ${propertyName}.`);
  }
}

/**
 * Re-calculates points for a previously scored race. This is necessary when the
 * total number of competitors in a series changes, affecting DNC/DNS scores.
 * It modifies the points for non-finishers directly on the provided race object.
 */
export function rescoreRacePoints(
  race: PublishedRace,
  newSeriesCompetitorCount: number,
  seriesType: SeriesScoringScheme
): void {
  // Determine the original ordering property. This assumes the race was scored correctly before.
  // We need to find the original handicap scheme from the series, which isn't passed in.
  // For now, we assume PY for handicap if not Level Rating.
  const results = race.results;
  const scheme = results.some(r => r.correctedTime !== r.elapsedTime) ? 'PY' : 'Level Rating';
  const orderingProperty = determineOrdering(race.type, scheme, results);

  // Re-assign points for finishers, as SCP penalties may change with series competitor count.
  assignPointsForFinishers(results, orderingProperty, seriesType, newSeriesCompetitorCount);
  
  // Re-assign points for non-finishers based on the new total competitor count.
  applyStaticRacePenalties(results, newSeriesCompetitorCount, seriesType);

  results.sort((a, b) => sortByPoints(a, b));
  calculateRanks(results);
}

function buildRaceResults(
  competitors: RaceCompetitor[]
): RaceResult[] {
  return competitors.map((comp) => {
    return {
      rank: comp.manualPosition || 0,
      boatClass: comp.boatClass,
      sailNumber: comp.sailNumber,
      helm: comp.helm,
      crew: comp.crew,
      laps: 0, // Will be set in calculateTimes
      handicap: comp.handicap,
      startTime: comp.startTime!,
      finishTime: comp.finishTime!,
      elapsedTime: 0,
      correctedTime: 0,
      points: 0,
      resultCode: comp.resultCode,
    };
  });
}

/**
 * Calculates elapsed and corrected times for each result.
 * from competitor timings
 */
function calculateTimes(results: RaceResult[], competitors: RaceCompetitor[], isAverageLap: boolean, scheme: HandicapSystem) {
  const maxLaps = competitors.reduce((max, comp) => (comp.numLaps > max) ? comp.numLaps : max, 0);

  for (let i = 0; i < results.length; i++) {
    const result = results[i];
    const competitor = competitors[i];

    result.laps = competitor.numLaps;
    result.elapsedTime = getElapsedTime(competitor, isAverageLap, maxLaps);
    result.correctedTime = calculateCorrectedTime(result.elapsedTime, result.handicap, scheme);
  }
}

/**
 * Calculates the elapsed time. For average lap races, 
 * the average lap time is scaled to the maximun number of laps completed
 * by any competitor 
*/
function getElapsedTime(comp: RaceCompetitor, isAverageLap: boolean, maxLaps: number): number {
  const finishTime = comp.finishTime;
  const compStartTime = comp.startTime;

  if (finishTime && compStartTime) {
    const diff = differenceInSeconds(finishTime.getTime(), compStartTime.getTime());

    if (diff < 0) {
      // Consider logging this error
      return 0;
    }

    const numLaps = comp.numLaps;
    if (isAverageLap && numLaps === 0) {
      // Consider logging this error
      return 0;
    }

    const elapsedTime = isAverageLap ? (diff / numLaps) * maxLaps : diff;
    return Math.round(elapsedTime);
  }
  return 0;
}

function calculateCorrectedTime(elapsedTime: number, handicap: number, scheme: HandicapSystem): number {
  if (elapsedTime === 0) {
    return 0;
  }
  switch (scheme) {
    case 'PY':
      return Math.round((elapsedTime * 1000.0) / handicap);
    case 'Level Rating':
      return elapsedTime; // No handicap correction
    case 'IRC':
    case 'Personal':
      throw new SailbrowserError(`${scheme} Not implemented`);
  }
}

/** Assigns points based on the competitor's ellapsed/corrected time.
 * Competitors with the same value for the given key (e.g., correctedTime or rank)
 * are awarded tied points.  
 * When multiple competitors are tied, the score is rounded to 1 decimal place. 
 * Scoring penalty codes are included and the scoting penatly applied 
 */
function assignPointsForFinishers(
  results: RaceResult[],
  key: keyof RaceResult,
  seriesType: SeriesScoringScheme,
  seriesCompetitorCount: number,
) {
  const finishers = results.filter((res) => isFinishedComp(res.resultCode));
  const resultsByValue = new Map<number, RaceResult[]>();

  // Group competitors by the value of the specified key
  for (const res of finishers) {
    const value = (res[key] as number) || 0;
    // A value of 0 is used for null for time/position, so we don't want to treat it as falsey.
    if (value === 0 && res.resultCode !== 'OK') continue;
    if (!resultsByValue.has(value)) resultsByValue.set(value, []);
    resultsByValue.get(value)!.push(res);
  }

  // Iterate over ordered list of values, calculating points for ties
  const sortedValues = Array.from(resultsByValue.keys()).sort((a, b) => a - b);
  let pos = 1.0;

  const startAreaCount = results.filter(r => isStartAreaComp(r.resultCode)).length;
  const dnfPoints = (seriesType === 'long' ? startAreaCount : seriesCompetitorCount) + 1;

  for (const value of sortedValues) {
    const resultsAtValue = resultsByValue.get(value)!;
    // The average points for a group of tied competitors is the average of the positions they would have taken.
    // For example, if 2 boats tie for 2nd, they take up positions 2 and 3, so they both get (2+3)/2 = 2.5 points.
    const avgPoints = pos - 1 + (resultsAtValue.length + 1) / 2.0;

    for (const res of resultsAtValue) {
      // Round ties to one decimal place 
      res.points = Math.round(avgPoints * 10) / 10;
      res.rank = pos;

      // If the competitor has a scoring penalty, apply it now.
      const algorithm = getShortAlgorithm(res.resultCode);
      if (algorithm === ResultCodeAlgorithm.scoringPenalty) {
        // RRS 44.3(c): Finish position + (20% * Boats Entered)
        const penalty = Math.round((dnfPoints * 0.2) * 10) / 10;
        // The penalty is capped at the DNF score.
        res.points = Math.min(res.points + penalty, dnfPoints);
      }
    }
    pos += resultsAtValue.length;
  }
}

/** Applies panalties that are not dependent on other 
 * results in the series. 
 */
function applyStaticRacePenalties(results: RaceResult[],
  seriesCompetitorCount: number,
  scheme: SeriesScoringScheme) {

  // 1. Calculate the number of boats that came to the start area for this specific race
  const startAreaCount = results.filter(r => isStartAreaComp(r.resultCode)).length;

  // This function should only apply penalties to non-finishers.
  // Finishers with penalties (like SCP) are handled in `assignPointsForFinishers`.
  const nonFinishers = results.filter(r => !isFinishedComp(r.resultCode));

  for (const result of nonFinishers) {
    // Determine which algorithm to use based on the scheme
    const algorithm = (scheme === 'long')
      ? getLongAlgorithm(result.resultCode)
      : getShortAlgorithm(result.resultCode);

    switch (algorithm) {
      case ResultCodeAlgorithm.compInSeries:
        result.points = seriesCompetitorCount + 1;
        break;
      case ResultCodeAlgorithm.compInStartArea:
        result.points = startAreaCount + 1;
        break;
      default:
        break;
    }
  }
}

/** 
 * Sorts by points.  Any boat that does not have any points yet 
 * assigned is sorted to the bottom.
 */
export function sortByPoints(a: RaceResult, b: RaceResult): number {
  return (a.points || 9999) - (b.points || 9999);
}

/**
 * Sorts results by a specified ordering property, ensuring that finishers always
 * appear before non-finishers.
 */
function sortByFinishingOrder(a: RaceResult, b: RaceResult, orderingProperty: keyof RaceResult): number {
  const aIsFinisher = isFinishedComp(a.resultCode);
  const bIsFinisher = isFinishedComp(b.resultCode);

  if (aIsFinisher && bIsFinisher) {
    return (a[orderingProperty] as number || 0) - (b[orderingProperty] as number || 0);
  } else if (aIsFinisher && !bIsFinisher) {
    return -1;
  } else if (!aIsFinisher && bIsFinisher) {
    return 1;
  } else {
    return 0; // Keep original order for non-finishers relative to each other
  }
}